# Project 09 -- Anthony Micco

**TA Help:** N/A
    
**Collaboration:** N/A

## Question 1

In [23]:
import pandas as pd
pd.options.mode.copy_on_write = True 

In [2]:
df = pd.read_csv("/anvil/projects/tdm/data/whin/weather.csv")

In [3]:
pd.set_option('display.max_columns',None)

In [4]:
df.head()

station_id  latitude  longitude             name      observation_time  \
0           1  40.93894  -86.47418  WHIN001-PULA001  2019-07-10T04:00:00Z   
1           1  40.93894  -86.47418  WHIN001-PULA001  2019-07-10T04:15:00Z   
2           1  40.93894  -86.47418  WHIN001-PULA001  2019-07-11T04:00:00Z   
3           1  40.93894  -86.47418  WHIN001-PULA001  2019-07-11T04:15:00Z   
4           1  40.93894  -86.47418  WHIN001-PULA001  2019-07-11T04:30:00Z   

   temperature  temperature_high  temperature_low  humidity  solar_radiation  \
0         70.0              71.0             70.0      83.0              NaN   
1         69.0              70.0             69.0      84.0              NaN   
2         76.0              77.0             76.0      76.0              NaN   
3         76.0              76.0             76.0      77.0              NaN   
4         76.0              76.0             76.0      77.0              NaN   

   solar_radiation_high  rain  rain_inches_last_hour  wind_speed_mph  \
0                   NaN   0.0                    0.0             0.0   
1                   NaN   0.0                    0.0             1.0   
2                   NaN   0.0                    0.0             2.0   
3                   NaN   0.0                    0.0             2.0   
4                   NaN   0.0                    0.0             2.0   

   wind_direction_degrees  wind_gust_speed_mph  wind_gust_direction_degrees  \
0                     NaN                  3.0                        247.5   
1                   247.5                  3.0                        247.5   
2                   202.5                  4.0                        202.5   
3                   202.5                  4.0                        202.5   
4                   225.0                  4.0                        202.5   

   pressure  soil_temp_1  soil_temp_2  soil_temp_3  soil_temp_4  soil_moist_1  \
0     30.05         77.0         78.0         76.0         74.0          24.0   
1     30.04         76.0         78.0         76.0         74.0          24.0   
2     29.89         80.0         80.0         78.0         75.0          31.0   
3     29.88         80.0         80.0         78.0         75.0          31.0   
4     29.88         80.0         80.0         78.0         75.0          32.0   

   soil_moist_2  soil_moist_3  soil_moist_4  
0          24.0          10.0           9.0  
1          25.0          10.0           9.0  
2          30.0          12.0          10.0  
3          31.0          12.0          10.0  
4          31.0          12.0          10.0

In [5]:
df.shape

(1000000, 26)

In [6]:
#Number of records using value counts
df["station_id"].value_counts()

station_id
1      71631
20     56917
147    45600
143    45593
146    45579
145    45509
144    45495
142    45395
149    44060
159    43563
153    43559
157    43534
155    42993
160    42810
156    42807
151    42806
164    38791
163    37908
166    30242
167    27463
168    18581
169    14905
172    14812
179    14772
176    14402
173    13648
175    13314
171    13311
Name: count, dtype: int64

In [7]:
#Number of records using group by
df.groupby("station_id").size()

station_id
1      71631
20     56917
142    45395
143    45593
144    45495
145    45509
146    45579
147    45600
149    44060
151    42806
153    43559
155    42993
156    42807
157    43534
159    43563
160    42810
163    37908
164    38791
166    30242
167    27463
168    18581
169    14905
171    13311
172    14812
173    13648
175    13314
176    14402
179    14772
dtype: int64

The difference between the two methodologies is that value_counts returns an object where the values of the columns are sorted in descending order with the element with the most entries is the first item. In contrast, group_by() method does not return an object, and sorts the elements in the grouped column rather than by the number of times they occur. I prefer using the value_counts() method because it is sorted by the number of times each element occurs rather than the column entries. 

## Question 2

In [8]:
#Determining the number of null values in each column
df.isna().sum()

station_id                          0
latitude                            0
longitude                           0
name                                0
observation_time                    0
temperature                      4176
temperature_high                 4123
temperature_low                  4123
humidity                         4162
solar_radiation                 48104
solar_radiation_high            12828
rain                                0
rain_inches_last_hour               0
wind_speed_mph                   4032
wind_direction_degrees          95545
wind_gust_speed_mph                 0
wind_gust_direction_degrees     94216
pressure                            0
soil_temp_1                    114043
soil_temp_2                    114041
soil_temp_3                    114037
soil_temp_4                    114036
soil_moist_1                   114124
soil_moist_2                   114170
soil_moist_3                   114174
soil_moist_4                   114150
dtype: int64

In [9]:
#Determining the total number of null values
df.isna().sum().sum()

1184084

In [10]:
#Dropping the null values and creating a new df
df_cleaned = df.dropna()

In [11]:
df_cleaned.head()

station_id  latitude  longitude             name      observation_time  \
12823           1  40.93894  -86.47418  WHIN001-PULA001  2019-11-21T19:15:00Z   
12824           1  40.93894  -86.47418  WHIN001-PULA001  2019-11-21T19:30:00Z   
12825           1  40.93894  -86.47418  WHIN001-PULA001  2019-11-21T19:45:00Z   
12826           1  40.93894  -86.47418  WHIN001-PULA001  2019-11-21T20:00:00Z   
12827           1  40.93894  -86.47418  WHIN001-PULA001  2019-11-21T20:15:00Z   

       temperature  temperature_high  temperature_low  humidity  \
12823         52.0              52.0             52.0      94.0   
12824         52.0              52.0             52.0      94.0   
12825         52.0              52.0             52.0      95.0   
12826         52.0              52.0             52.0      94.0   
12827         52.0              52.0             52.0      95.0   

       solar_radiation  solar_radiation_high  rain  rain_inches_last_hour  \
12823             31.0                  42.0   0.0                    0.0   
12824             24.0                  32.0   0.0                    0.0   
12825             23.0                  26.0   0.0                    0.0   
12826             29.0                  46.0   0.0                    0.0   
12827             17.0                  25.0   0.0                    0.0   

       wind_speed_mph  wind_direction_degrees  wind_gust_speed_mph  \
12823            19.0                   225.0                 29.0   
12824            20.0                   225.0                 29.0   
12825            20.0                   225.0                 27.0   
12826            19.0                   225.0                 27.0   
12827            19.0                   225.0                 27.0   

       wind_gust_direction_degrees  pressure  soil_temp_1  soil_temp_2  \
12823                        225.0    29.805         46.0         44.0   
12824                        225.0    29.805         46.0         44.0   
12825                        225.0    29.810         46.0         44.0   
12826                        225.0    29.813         46.0         44.0   
12827                        225.0    29.814         46.0         44.0   

       soil_temp_3  soil_temp_4  soil_moist_1  soil_moist_2  soil_moist_3  \
12823         44.0         44.0           7.0          14.0          11.0   
12824         44.0         44.0           6.0          14.0          11.0   
12825         44.0         44.0           6.0          14.0          11.0   
12826         44.0         44.0           6.0          14.0          11.0   
12827         44.0         44.0           6.0          14.0          11.0   

       soil_moist_4  
12823           9.0  
12824           9.0  
12825           9.0  
12826           9.0  
12827           9.0

In [12]:
df_cleaned.isna().sum()

station_id                     0
latitude                       0
longitude                      0
name                           0
observation_time               0
temperature                    0
temperature_high               0
temperature_low                0
humidity                       0
solar_radiation                0
solar_radiation_high           0
rain                           0
rain_inches_last_hour          0
wind_speed_mph                 0
wind_direction_degrees         0
wind_gust_speed_mph            0
wind_gust_direction_degrees    0
pressure                       0
soil_temp_1                    0
soil_temp_2                    0
soil_temp_3                    0
soil_temp_4                    0
soil_moist_1                   0
soil_moist_2                   0
soil_moist_3                   0
soil_moist_4                   0
dtype: int64

In [13]:
df_cleaned.shape

(767725, 26)

There are 767725 rows and 26 columns in df_cleaned.

## Question 3

In [14]:
#Creating df with no null temp values
tempDF = df.dropna(subset=["temperature"])

In [15]:
tempDF.head()

station_id  latitude  longitude             name      observation_time  \
0           1  40.93894  -86.47418  WHIN001-PULA001  2019-07-10T04:00:00Z   
1           1  40.93894  -86.47418  WHIN001-PULA001  2019-07-10T04:15:00Z   
2           1  40.93894  -86.47418  WHIN001-PULA001  2019-07-11T04:00:00Z   
3           1  40.93894  -86.47418  WHIN001-PULA001  2019-07-11T04:15:00Z   
4           1  40.93894  -86.47418  WHIN001-PULA001  2019-07-11T04:30:00Z   

   temperature  temperature_high  temperature_low  humidity  solar_radiation  \
0         70.0              71.0             70.0      83.0              NaN   
1         69.0              70.0             69.0      84.0              NaN   
2         76.0              77.0             76.0      76.0              NaN   
3         76.0              76.0             76.0      77.0              NaN   
4         76.0              76.0             76.0      77.0              NaN   

   solar_radiation_high  rain  rain_inches_last_hour  wind_speed_mph  \
0                   NaN   0.0                    0.0             0.0   
1                   NaN   0.0                    0.0             1.0   
2                   NaN   0.0                    0.0             2.0   
3                   NaN   0.0                    0.0             2.0   
4                   NaN   0.0                    0.0             2.0   

   wind_direction_degrees  wind_gust_speed_mph  wind_gust_direction_degrees  \
0                     NaN                  3.0                        247.5   
1                   247.5                  3.0                        247.5   
2                   202.5                  4.0                        202.5   
3                   202.5                  4.0                        202.5   
4                   225.0                  4.0                        202.5   

   pressure  soil_temp_1  soil_temp_2  soil_temp_3  soil_temp_4  soil_moist_1  \
0     30.05         77.0         78.0         76.0         74.0          24.0   
1     30.04         76.0         78.0         76.0         74.0          24.0   
2     29.89         80.0         80.0         78.0         75.0          31.0   
3     29.88         80.0         80.0         78.0         75.0          31.0   
4     29.88         80.0         80.0         78.0         75.0          32.0   

   soil_moist_2  soil_moist_3  soil_moist_4  
0          24.0          10.0           9.0  
1          25.0          10.0           9.0  
2          30.0          12.0          10.0  
3          31.0          12.0          10.0  
4          31.0          12.0          10.0

In [24]:
#combining the lat and long columns
tempDF["location"] = df["latitude"].astype(str) + ", " + df["longitude"].astype(str)

In [26]:
tempDF[["location"]]

location
0                       40.93894, -86.47418
1                       40.93894, -86.47418
2                       40.93894, -86.47418
3                       40.93894, -86.47418
4                       40.93894, -86.47418
...                                     ...
999995  40.2967994865571, -87.3902853950858
999996  40.2967994865571, -87.3902853950858
999997  40.2967994865571, -87.3902853950858
999998  40.2967994865571, -87.3902853950858
999999  40.2967994865571, -87.3902853950858

[995824 rows x 1 columns]

In [45]:
#Grouping the data by location and finding average temp
tempDF.groupby("location")["temperature"].mean().sort_values()

location
40.38116, -86.40269                    47.765668
40.2709571472446, -87.1486035394257    49.803342
40.93894, -86.47418                    49.883742
40.42124, -86.84642                    50.742410
40.48079, -87.20682                    51.730032
41.01865, -86.7101                     53.836086
40.16179, -87.35246                    54.052897
40.98224, -86.38542                    54.206106
40.58629, -87.43654                    54.676701
40.78049, -86.89576                    54.791890
40.37634, -86.59591                    54.794076
40.5485, -87.12477                     54.898998
40.84436, -86.18173                    54.942505
40.53722, -86.95342                    55.070667
40.43134, -86.53464                    55.081819
40.38539, -87.51034                    55.115662
40.51432, -86.45856                    55.183031
40.97061, -86.35304                    55.292530
40.59057, -86.39155                    55.307808
40.70178, -86.70649                    55.482287
40.10483, -

## Question 4

In [51]:
# creating a function to question 3 work
def drop_null_find_avg_temp(dataframe):
    
    #Creating a new df and dropping the null temperature values
    tmpDF =  dataframe.dropna(subset=["temperature"])
    
    #Creating a new location column by combining lat and long columns
    tmpDF["location"] = dataframe["latitude"].astype(str) + ", " + dataframe["longitude"].astype(str)
    
    #Return the average temperature at each location
    return tmpDF.groupby("location")["temperature"].mean().sort_values()

In [52]:
drop_null_find_avg_temp(df)

location
40.38116, -86.40269                    47.765668
40.2709571472446, -87.1486035394257    49.803342
40.93894, -86.47418                    49.883742
40.42124, -86.84642                    50.742410
40.48079, -87.20682                    51.730032
41.01865, -86.7101                     53.836086
40.16179, -87.35246                    54.052897
40.98224, -86.38542                    54.206106
40.58629, -87.43654                    54.676701
40.78049, -86.89576                    54.791890
40.37634, -86.59591                    54.794076
40.5485, -87.12477                     54.898998
40.84436, -86.18173                    54.942505
40.53722, -86.95342                    55.070667
40.43134, -86.53464                    55.081819
40.38539, -87.51034                    55.115662
40.51432, -86.45856                    55.183031
40.97061, -86.35304                    55.292530
40.59057, -86.39155                    55.307808
40.70178, -86.70649                    55.482287
40.10483, -

## Question 5

In [48]:
#Creating df with no null wind gust speed values
windDF =  df.dropna(subset=["wind_gust_speed_mph"])

In [34]:
windDF.head()

station_id  latitude  longitude             name      observation_time  \
0           1  40.93894  -86.47418  WHIN001-PULA001  2019-07-10T04:00:00Z   
1           1  40.93894  -86.47418  WHIN001-PULA001  2019-07-10T04:15:00Z   
2           1  40.93894  -86.47418  WHIN001-PULA001  2019-07-11T04:00:00Z   
3           1  40.93894  -86.47418  WHIN001-PULA001  2019-07-11T04:15:00Z   
4           1  40.93894  -86.47418  WHIN001-PULA001  2019-07-11T04:30:00Z   

   temperature  temperature_high  temperature_low  humidity  solar_radiation  \
0         70.0              71.0             70.0      83.0              NaN   
1         69.0              70.0             69.0      84.0              NaN   
2         76.0              77.0             76.0      76.0              NaN   
3         76.0              76.0             76.0      77.0              NaN   
4         76.0              76.0             76.0      77.0              NaN   

   solar_radiation_high  rain  rain_inches_last_hour  wind_speed_mph  \
0                   NaN   0.0                    0.0             0.0   
1                   NaN   0.0                    0.0             1.0   
2                   NaN   0.0                    0.0             2.0   
3                   NaN   0.0                    0.0             2.0   
4                   NaN   0.0                    0.0             2.0   

   wind_direction_degrees  wind_gust_speed_mph  wind_gust_direction_degrees  \
0                     NaN                  3.0                        247.5   
1                   247.5                  3.0                        247.5   
2                   202.5                  4.0                        202.5   
3                   202.5                  4.0                        202.5   
4                   225.0                  4.0                        202.5   

   pressure  soil_temp_1  soil_temp_2  soil_temp_3  soil_temp_4  soil_moist_1  \
0     30.05         77.0         78.0         76.0         74.0          24.0   
1     30.04         76.0         78.0         76.0         74.0          24.0   
2     29.89         80.0         80.0         78.0         75.0          31.0   
3     29.88         80.0         80.0         78.0         75.0          31.0   
4     29.88         80.0         80.0         78.0         75.0          32.0   

   soil_moist_2  soil_moist_3  soil_moist_4             location  
0          24.0          10.0           9.0  40.93894, -86.47418  
1          25.0          10.0           9.0  40.93894, -86.47418  
2          30.0          12.0          10.0  40.93894, -86.47418  
3          31.0          12.0          10.0  40.93894, -86.47418  
4          31.0          12.0          10.0  40.93894, -86.47418

In [53]:
#combining the lat and long columns
windDF["location"] = df["latitude"].astype(str) + ", " + df["longitude"].astype(str)

In [54]:
windDF[["location"]]

location
0                       40.93894, -86.47418
1                       40.93894, -86.47418
2                       40.93894, -86.47418
3                       40.93894, -86.47418
4                       40.93894, -86.47418
...                                     ...
999995  40.2967994865571, -87.3902853950858
999996  40.2967994865571, -87.3902853950858
999997  40.2967994865571, -87.3902853950858
999998  40.2967994865571, -87.3902853950858
999999  40.2967994865571, -87.3902853950858

[1000000 rows x 1 columns]

In [49]:
#Grouping the data by location and finding average wind gust speed
windDF.groupby("location")["wind_gust_speed_mph"].mean().sort_values()

location
40.42124, -86.84642                     5.973414
40.3840069271445, -87.31664031744       6.184488
40.48079, -87.20682                     6.290135
40.3861140862113, -87.1012964844704     7.812415
41.01865, -86.7101                      8.975526
40.16179, -87.35246                     9.026960
40.70178, -86.70649                     9.843719
40.93894, -86.47418                     9.944437
40.2709571472446, -87.1486035394257     9.993939
40.59057, -86.39155                    10.055651
40.1492432088387, -86.7371410131454    10.136323
40.30156, -87.48248                    10.146300
40.37634, -86.59591                    10.174628
40.51432, -86.45856                    10.181839
40.10483, -86.86619                    10.186783
40.84436, -86.18173                    10.285053
40.5485, -87.12477                     10.516033
40.97061, -86.35304                    10.564448
40.78049, -86.89576                    10.852191
40.98224, -86.38542                    10.964359
40.38539, -

In [55]:
#Creating function for dropping null values and finding average wind gust speed
def drop_null_find_avg_wind_gust(dataframe):
    
    #Creating a new df and dropping the null wind gust speed values
    tmp_windDF =  dataframe.dropna(subset=["wind_gust_speed_mph"])
    
    #Creating a new location column by combining lat and long columns
    tmp_windDF["location"] = dataframe["latitude"].astype(str) + ", " + dataframe["longitude"].astype(str)
    
    #Return the average wind gust speed at each location
    return tmp_windDF.groupby("location")["wind_gust_speed_mph"].mean().sort_values()

In [56]:
drop_null_find_avg_wind_gust(df)

location
40.42124, -86.84642                     5.973414
40.3840069271445, -87.31664031744       6.184488
40.48079, -87.20682                     6.290135
40.3861140862113, -87.1012964844704     7.812415
41.01865, -86.7101                      8.975526
40.16179, -87.35246                     9.026960
40.70178, -86.70649                     9.843719
40.93894, -86.47418                     9.944437
40.2709571472446, -87.1486035394257     9.993939
40.59057, -86.39155                    10.055651
40.1492432088387, -86.7371410131454    10.136323
40.30156, -87.48248                    10.146300
40.37634, -86.59591                    10.174628
40.51432, -86.45856                    10.181839
40.10483, -86.86619                    10.186783
40.84436, -86.18173                    10.285053
40.5485, -87.12477                     10.516033
40.97061, -86.35304                    10.564448
40.78049, -86.89576                    10.852191
40.98224, -86.38542                    10.964359
40.38539, -

The largest average wind gust speed is 12.850781 mph for the location 40.38116, -86.40269.

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.